code/datums/alert.dm

```diff
@@ -0,0 +1,181 @@
+///Represents a single source of alerts, one alert handler will only ever count for one alert per listener
+/datum/alert_handler
+	///A list of alert type -> list of area names we currently have alerts in
+	var/list/sent_alerts = list()
+	///Our source atom
+	var/atom/source_atom
+
+/datum/alert_handler/New(atom/_source_atom)
+	if(istype(_source_atom))
+		source_atom = _source_atom
+	else
+		var/source_type = ""
+		if(istype(_source_atom, /datum))
+			source_type = _source_atom.type
+		stack_trace("a non atom was passed into alert_handler! [_source_atom] [source_type]")
+	return ..()
+
+/datum/alert_handler/Destroy()
+	for(var/alert_type in sent_alerts)
+		for(var/area/area_to_clear in sent_alerts[alert_type])
+			//Yeet all connected alerts
+			clear_alert_from_area(alert_type, area_to_clear)
+	source_atom = null
+	return ..()
+
+///Sends an alert to any interested things, does some checks to prevent unneeded work
+///Important to note is that source_atom is not held as a ref, we're used as a proxy to prevent hard deletes
+///optional_camera should only be used when you have one camera you want to pass along to alert listeners, most of the time you should have no use for it
+/datum/alert_handler/proc/send_alert(alert_type, atom/use_as_source_atom = source_atom, optional_camera)
```

Comment:
Aren't default arguments that aren't constants like, bad for some reason?

Can you change this to just `use_as_source_atom = use_as_source_atom || source_atom`?

Reply:
sure